lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -5932,6 +5971,34 @@ LLVMToSPIRVBase::transBuiltinToInstWithoutDecoration(Op OC, CallInst *CI,
                                    transValue(Image, BB),
                                    transValue(Sampler, BB), BB);
   }
+  case OpImageRead:
+  case OpImageSampleExplicitLod:
+  case OpImageWrite: {
+    // Image Op needs handling of SignExtend or ZeroExtend image operands.
+    auto Args = getArguments(CI);
+    SPIRVType *SPRetTy =
+        CI->getType()->isVoidTy() ? nullptr : transScavengedType(CI);
+    auto *SPI = SPIRVInstTemplateBase::create(OC);
+    std::vector<SPIRVWord> SPArgs;
+    for (size_t I = 0, E = Args.size(); I != E; ++I) {
+      if (Args[I]->getType()->isPointerTy()) {
+        Value *Pointee = Args[I]->stripPointerCasts();
+        (void)Pointee;
+        assert((Pointee == Args[I] || !isa<Function>(Pointee)) &&
+               "Illegal use of a function pointer type");
+      }
+      SPArgs.push_back(SPI->isOperandLiteral(I)
+                           ? cast<ConstantInt>(Args[I])->getZExtValue()
+                           : transValue(Args[I], BB)->getId());
+    }
+    if (BM->isAllowedToUseVersion(VersionNumber::SPIRV_1_4)) {
+      if (Args.size() <= getImageOperandsIndex(OC))
+        if (int SignZeroExt = getImageSignZeroExt(CI->getCalledFunction()))
+          SPArgs.push_back(SignZeroExt);
```

Comment:
Basically this line of code has triggered the original question about necessity of the patch.
We can reach this code in 2 cases:
1. As you have described after we have converted LLVM IR with OpenCL flavor to SPIR-V friendly LLVM IR, like this there should be not issues with this code;
2. Just be translating hand-written SPIR-V friendly LLVM IR, like Intel's SYCL device headers do.

If I read the code correctly in this case there will be an issue if we have encoded other operands in the image instruction call - we will add yet another operand, while we should be modifying existing Image Operands parameter by applying the appropriate mask, like we have in the input IR:
`%image = call <4 x i32> @_Z17__spirv_ImageReadIDv4_j14ocl_image3d_roDv4_iET_T0_T1_(target("spirv.Image", void, 2, 0, 0, 0, 0, 0, 0) undef, <4 x i32> zeroinitializer, /* image operand bias*/ i32 0x1, /*literal float value of bias*/ float 2.0f)`
no we want to add zero extend mask, the result is needed to be `...  /* image operand bias + zero extend mask*/ i32 0x1 | 0x2000, /*literal float value of bias*/ float 2.0f) ` , but instead (if I read the code correctly) we would get `...  /* image operand bias + zero extend mask*/ i32 0x1, /*literal float value of bias*/ float 2.0f, i32 0x2000) ` which is incorrect. I might be missing something here, but even if so, may I ask you to add a test proving, that I'm wrong.

Reply:
> if we have encoded other operands in the image instruction call - we will add yet another operand

New ImageOperands won't be added if there is already an ImageOperands, see the check at https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/2286/files/c7826f11bc4730727d144f65845730fadb8a48c6#diff-ba003da77a80df1db76fb378a873a80e758591d72bf82a46fc90069b2a947990R5995

> while we should be modifying existing Image Operands parameter by applying the appropriate mask

Thank you for pointing out this, which was missing in previous commits. I've just added this part. Test is https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/2286/files#diff-52ffa59d3f317359a71e063620bf6052e847aebd24cac407a69a2f2398a2699cR31